src/MergeContext.js

```diff
@@ -198,19 +200,157 @@ class StatusChecks
     }
 }
 
-// Common methods/fields for MergeInitiator and MergeFinalizer.
-// Do not create directly.
-class MergeContext {
+// Cached label state
+class Label
+{
+    constructor(name, on, old) {
+        this.name = name;
+        this.on = on; // current state
+        // whether this label was received from GitHub or created by the bot
+        this.old = old;
+    }
+
+    toString() {
+        return this.name + "(on: " + this.on + "; old: " + this.old + ")";
+    }
+}
+
+// Keeps labels state (received from GitHub), caches labels modifications
+// and applies them back to GitHub
+class Labels
+{
+    // labels parameter keeps label array received from GitHub
+    constructor(labels, prNum) {
+        this._prNum = prNum;
+        this._labels = [];
+        for (let label of labels)
+            this._doCache(label.name, true, true);
+    }
+
+    _doCache(name, on, old) {
+        const label = this._labels.find(lbl => lbl.name === name);
+        if (label)
+            label.on = on;
```

Comment:
I left this method, renaming to modify(): it helps to avoid duplication between add() and remove().

Reply:
AFAICT, the _only_ duplication is the complex .find() call (which can and should be converted into a simple this.find() method call. 

Please split and move _modify() logic into the corresponding methods that currently convert meaningful "add" and "remove" names into a rather awkward mustBeSet boolean parameter. It will make the code a lot easier to comprehend.

The only reason to keep _modify() (as a public modify() wrapper that simply calls the other two methods) is to simplify _labelFailedDescription(). Your call whether to do that or not.